### DISCUSSION TITLE
Any way to print PDF into memory, instead of to a file?

### DISCUSSION THREAD

**Discussion by danya02** [cefsharp/CefSharp#4877] [Q&A] [answered]
I'm making a web server that returns PDFs rendered from HTML templates. Right now I have to create a temporary file, pass its name to the `PrintToPdfAsync` method, and then read the contents into memory before sending them out to the user and deleting the file. I'd like to skip the step of going to the filesystem, and instead get the PDF bytes directly.

From what I can tell, it seems that [CEF's printToPDF method](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF) returns a base64-encoded PDF bytes, or alternatively an experimental `IO.StreamHandle`. Could this be used for my goal?

**Comment by campersau** [comment]
You can try out the dev tools client. Here is an example for screenshots but the CDP  printToPdf API you mentioned is also available:
https://github.com/cefsharp/CefSharp/wiki/General-Usage#screenshots

**Reply by danya02** [reply]
Thanks for the quick reply! I found that too, I was actually using the `IBrowserHost` method. I tried using the `DevToolsClient`, but it replies with:

```
CefSharp.DevTools.DevToolsClientException: DevTools Client Error :'Page.printToPDF' wasn't found
```

I think this issue is related: https://groups.google.com/a/chromium.org/g/chromium-extensions/c/gZ-jLD1bgCY

**Reply by amaitland** [reply]
What version are you using? 

> CefSharp.DevTools.DevToolsClientException: DevTools Client Error :'Page.printToPDF' wasn't found

Previously this was only supported in headless implementation, that should have been fixed in recent versions.

https://issues.chromium.org/issues/40533923

**Reply by danya02** [reply]
> What version are you using?

```
    <TargetFramework>netcoreapp3.1</TargetFramework>
...
    <PackageReference Include="CefSharp.Common.NetCore" Version="125.0.210" />
    <PackageReference Include="CefSharp.OffScreen.NetCore" Version="125.0.210" />
```

> only supported in headless implementation

That's what I'm trying to use, though. I can't use the headful browser on a web server anyway.

**Reply by amaitland** [reply, marked as answer]
OffScreen and headless are different things. `CefSharp.OffScreen` is technically a non-headless browser even though it renders to a bitmap.

Try switching to puppeteer sharp, you should be able to get the PDF in memory.
